### PR TITLE
Feature/#600 add assignments

### LINKED ISSUE #600: Add assignments to geli
# User Story

### As a: 
Student/Teacher

### I want:
a possible infrastructure for uploading assignments.

## Acceptance criteria:
- [ ] Any group can see there assignment, unlike moodle only one person can see the uploaded file
- [x] Assignments have a duty date, this date should be changeable 
- [x] For Labs there should be a duplicate function
- [x] The teacher can grade the assignment(ok/not ok) and can request changes
- [ ] Assignments can be multiple files, so that the student can swap file easily instead of packing a zip file every time

## Estimated:
- 8 x 4hr
------
_Please tag this issue if you are sure to which tag(s) it belongs._

### PR DESCRIPTION
**Pull request by lukas-schardt**
## Description:

Closes #600

## Improvements
- 

## Known Issues:
_NONE_

### CODE REVIEW

**Comment by lukas-schardt** [comment]
Anyone who wants to review this pull request is welcome to do so. It is still WIP because of missing tests.

**Comment by kesselb** [comment]
Peek 2019-03-22 22-34

**Comment by kesselb** [comment]
Some conflicts with https://github.com/geli-lms/geli/pull/1191 :disappointed:

**Comment by torss** [comment]
> Some conflicts with #1191 😞

Hm the `UnitController` in here would also conflict with some changes necessary for #1196 because of the `File` usage. (Although I'm not sure whether #1196 will be completed before Tuesday anyway.)

**Comment by kesselb** [comment]
It's not possible to download a assignment unit. Could you exclude assignment units from download feature? (we can add some kind of export for assigment units later)

**Comment by lukas-schardt** [comment]
@kesselb Thanks. Would've missed it. Will do.

File: app/webFrontend/src/app/unit/assignment-unit/assignment-unit.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { AssignmentUnitComponent } from './assignment-unit.component';
```
Comment (kesselb) [inline app/webFrontend/src/app/unit/assignment-unit/assignment-unit.component.spec.ts]: Is this test working? If not please delete it ;)

File: app/webFrontend/src/app/unit/assignment-unit/assignment-unit.component.ts [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  }
+
+  public submitAssignment() {
+    this.assignment.submitted = true;
+    this.assignmentService.updateAssignment(this.assignment, this.assignmentUnit._id.toString());
+  }
+
+  public async downloadAllAssignments() {
+    try {
+      this.disableDownloadButton = true;
```
Comment (kesselb) [inline app/webFrontend/src/app/unit/assignment-unit/assignment-unit.component.ts]:
```
this.disableDownloadButton = true;
```

Could you move this before try-catch?

File: app/webFrontend/src/app/unit/assignment-unit/assignment-unit.component.ts [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+      saveAs(response.body, this.assignmentUnit.name + '.zip');
+    } catch (err) {
+      this.snackBar.openLong('unit.assignment.errorDownload');
+    }
+    this.disableDownloadButton = false;
+  }
+
+  async downloadSingleAssignment(assignment) {
+    try {
+      this.disableDownloadButton = true;
```
Comment (kesselb) [inline app/webFrontend/src/app/unit/assignment-unit/assignment-unit.component.ts]:
```
this.disableDownloadButton = true;
```

Could you move this before try-catch?

File: app/webFrontend/src/app/unit/unit-form/assignment-unit-form/assignment-unit-form.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { AssignmentUnitFormComponent } from './assignment-unit-form.component';
+
+describe('AssignmentUnitFormComponent', () => {
```
Comment (kesselb) [inline app/webFrontend/src/app/unit/unit-form/assignment-unit-form/assignment-unit-form.component.spec.ts]: Is this test working? If not please delete it.

File: app/webFrontend/src/assets/i18n/de.json [hunk 5]
```diff
@@ -87,7 +87,8 @@
     "discard": "Verwerfen",
     "dismiss": "Schließen",
     "documents": "Dokumente",
-    "download": "Download (Beta)",
+    "download": "Herunterladen",
```
Comment (kesselb) [inline app/webFrontend/src/assets/i18n/de.json]: "download" is not used anymore (we use "downloadBeta" until most of the issues are fixed)

File: app/webFrontend/src/assets/i18n/en.json [hunk 6]
```diff
@@ -86,7 +86,8 @@
     "description": "Description",
     "discard": "Discard",
     "dismiss": "Dismiss",
-    "download": "Download (Beta)",
+    "download": "Download",
```
Comment (kesselb) [inline app/webFrontend/src/assets/i18n/en.json]: "download" is not used anymore (we use "downloadBeta" until most of the issues are fixed)

File: app/webFrontend/src/app/unit/assignment-unit/assignment-unit.component.html [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        </span>
+        </td>
+        <td class="{{!assignment.submitted ? 'not-submitted' : ''}}">
+          <mat-expansion-panel>
+            <mat-expansion-panel-header>
+              <mat-panel-title>
+                Dateien anzeigen
+              </mat-panel-title>
+              <mat-panel-description>
+                {{ 'unit.assignment.numberOfFiles' | translate:{filesCount: assignment.files.length} }}
```
Comment (kesselb) [inline app/webFrontend/src/app/unit/assignment-unit/assignment-unit.component.html]: You could try to use https://angular.io/guide/i18n#pluralization here

File: api/gulpfile.js [hunk 8]
```diff
@@ -145,6 +145,7 @@ gulp.task(BUILD_DEV, function (callback) {
 });
 */
 
+
```
Comment (kesselb) [inline api/gulpfile.js]: Could you revert this?

File: api/src/models/units/Unit.ts [hunk 9]
```diff
@@ -71,6 +72,13 @@ const unitSchema = new mongoose.Schema({
           ret._course = ret._course.toString();
         }
 
+        if (ret.assignments) {
+          ret.assigments = ret.assignments.map((assignment: any) => {
+            assignment._id = '';
```
Comment (kesselb) [inline api/src/models/units/Unit.ts]: Why do you unset the id here?

File: app/webFrontend/src/assets/i18n/de.json (hunk 5 above)
Comment (lukas-schardt) [reply app/webFrontend/src/assets/i18n/de.json]: It is used in the teacher view of the assignment units.

File: api/src/models/units/Unit.ts (hunk 9 above)
Comment (lukas-schardt) [reply api/src/models/units/Unit.ts]: I have no idea... I think I was trying to achieve something else and this is a remnant. Removing it

File: app/webFrontend/src/app/unit/assignment-unit/assignment-unit.component.html (hunk 7 above)
Comment (lukas-schardt) [reply app/webFrontend/src/app/unit/assignment-unit/assignment-unit.component.html]: Looks quite a bit complicated. I would like to keep it as it is, as it is more readable.

File: app/webFrontend/src/assets/i18n/de.json (hunk 5 above)
Comment (kesselb) [reply app/webFrontend/src/assets/i18n/de.json]: Ok

**Comment by kesselb** [comment]
Conflicts :disappointed: :see_no_evil:

**Comment by PatrickSkowronek** [comment]
@lukas-schardt @kesselb good work! I think this feature will not run smoothly on Prod. I know you test a lot, but this is the feature that cloud make it usable for other courses then FWE. So please be 100% sure that it will run good. Unfortunately I have no time to test it.

**Comment by dboschm** [comment]
The teachers view of the assignment unit is not responsive.
grafik

**Comment by lukas-schardt** [comment]
@dboschm Thanks. Fixed.

**Pull request merged**